### PR TITLE
Handling Permissions denials #2692

### PR DESCRIPTION
**Pull request by cnevinc**
I have a strange situation here.
Maybe you've already seen it?

See https://github.com/mozilla-tw/Rocket/compare/nevin/permission?expand=1#diff-29976a057276bb0f502b51b61ba7dfe7R155

### CODE REVIEW

File: app/src/main/java/org/mozilla/focus/permission/PermissionHandler.java [hunk 1]
```diff
@@ -141,13 +146,18 @@ public void onRequestPermissionsResult(Context context, int requestCode, @NonNul
...
         if (requestCode == actionId) {
-            setPermissionAsked(context, permissions[0]);
+
             if (grantResults.length > 0 && grantResults[0] == PackageManager.PERMISSION_GRANTED) {
                 permissionHandle.doActionGranted(permission, actionId, params);
                 clearAction();
             } else {
-                showSnackBar();
+                showToast();
+                // I'm not sure why it's so. This method already on Main thread.
```
Comment (mTwTm) [inline app/src/main/java/org/mozilla/focus/permission/PermissionHandler.java]: You need to do permissionNotGranted() here, note that we do call this in showSnackBar() but not in showToast(). Sorry for writing bad interfaces :P

File: app/src/main/java/org/mozilla/focus/fragment/BrowserFragment.java [hunk 2]
```diff
@@ -334,6 +335,11 @@ public void requestPermissions(int actionId) {
                         throw new IllegalArgumentException("Unknown Action");
                 }
             }
+
+            @Override
+            public void permissionDeniedToast() {
```
Comment (mTwTm) [inline app/src/main/java/org/mozilla/focus/fragment/BrowserFragment.java]: We have different permissions, location and storage. Pass actionId to this interface, see Snackbar makeAskAgainSnackBar(int actionId);

File: app/src/main/java/org/mozilla/focus/permission/PermissionHandler.java [hunk 3]
```diff
@@ -141,13 +146,18 @@ public void onRequestPermissionsResult(Context context, int requestCode, @NonNul
...
         }
         if (requestCode == actionId) {
-            setPermissionAsked(context, permissions[0]);
+
             if (grantResults.length > 0 && grantResults[0] == PackageManager.PERMISSION_GRANTED) {
                 permissionHandle.doActionGranted(permission, actionId, params);
                 clearAction();
             } else {
-                showSnackBar();
+                showToast();
```
Comment (mTwTm) [inline app/src/main/java/org/mozilla/focus/permission/PermissionHandler.java]: Tried if we don't reject, it's still incorrect. Maybe we should first show the dialog first and only if the user wants to give the site permission then we ask for device permission. This is how Firefox and Chrome handle this.

File: app/src/main/java/org/mozilla/focus/permission/PermissionHandler.java [hunk 4]
```diff
@@ -141,13 +146,18 @@ public void onRequestPermissionsResult(Context context, int requestCode, @NonNul
...
+                // I'm not sure why it's so. This method already on Main thread.
+                // But if I don't do this, webview will keeps requesting for permission.
+                // See https://github.com/mozilla-tw/Rocket/blob/765f6a1ddbc2b9058813e930f63c62a9797c5fa0/app/src/webkit/java/org/mozilla/focus/webkit/FocusWebChromeClient.java#L126
+                ThreadUtils.postToMainThread(this::permissionNotGranted);
             }
+            setPermissionAsked(context, permissions[0]);
```
Comment (mTwTm) [inline app/src/main/java/org/mozilla/focus/permission/PermissionHandler.java]: Do we need this?

File: app/src/main/java/org/mozilla/focus/fragment/BrowserFragment.java (hunk 2 above)
Comment (cnevinc) [reply app/src/main/java/org/mozilla/focus/fragment/BrowserFragment.java]: fixed

File: app/src/main/java/org/mozilla/focus/fragment/BrowserFragment.java [hunk 5]
```diff
@@ -336,6 +338,11 @@ public void requestPermissions(int actionId) {
                         throw new IllegalArgumentException("Unknown Action");
                 }
             }
+
+            @Override
+            public void permissionDeniedToast(int actionId) {
+                Toast.makeText(getContext(), getAskAgainSnackBarString(actionId ), Toast.LENGTH_LONG).show();
```
Comment (mTwTm) [inline app/src/main/java/org/mozilla/focus/fragment/BrowserFragment.java]: I think this is not what the spec describes?

Comment (cnevinc) [reply app/src/main/java/org/mozilla/focus/fragment/BrowserFragment.java]: fixed

File: app/src/main/java/org/mozilla/focus/fragment/BrowserFragment.java [hunk 6]
```diff
@@ -336,6 +348,11 @@ public void requestPermissions(int actionId) {
                         throw new IllegalArgumentException("Unknown Action");
                 }
             }
+
+            @Override
+            public void permissionDeniedToast(int actionId) {
+                Toast.makeText(getContext(), getPermissionDeniedToastString(actionId ), Toast.LENGTH_LONG).show();
```
Comment (mTwTm) [inline app/src/main/java/org/mozilla/focus/fragment/BrowserFragment.java]: NIT: extra space.

File: app/src/main/res/values/strings.xml [hunk 7]
```diff
@@ -250,6 +250,8 @@
     <string name="permission_toast_location">Location not available. Give permission to access your location.</string>
     <string name="permission_dialog_msg_storage">Missing permissions. Give Firefox Rocket permission to access your location.</string>
     <string name="permission_toast_storage">Storage not available. Give permission to store photos and files on your device.</string>
+    <string name="location_permission_is_denied">Location permission is denied.</string>
```
Comment (mTwTm) [inline app/src/main/res/values/strings.xml]: We already have: permission_toast_location_deny

File: app/src/main/res/values/strings.xml [hunk 8]
```diff
@@ -250,6 +250,8 @@
...
+    <string name="storage_permission_is_denied">Storage permission is denied.</string>
```
Comment (mTwTm) [inline app/src/main/res/values/strings.xml]: Same here.

**Pull request merged**